extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -466,6 +466,17 @@ && isRedirectFromProvider(context, configContext)) {
                         codeFlowParams.append(AMP).append(OidcConstants.TOKEN_SCOPE).append(EQ)
                                 .append(OidcCommonUtils.urlEncode(String.join(" ", scopes)));
 
+                        if (!configContext.oidcConfig.getAuthentication().forwardParams.isEmpty()) {
+                            MultiMap requestQueryParams = context.queryParams();
+                            for (String forwardedParam : configContext.oidcConfig.getAuthentication().forwardParams.get()) {
+                                if (requestQueryParams.contains(forwardedParam)) {
+                                    codeFlowParams.append(AMP).append(forwardedParam).append(EQ)
+                                            .append(OidcCommonUtils
+                                                    .urlEncode(String.join(" ", requestQueryParams.getAll(forwardedParam))));
```

Comment:
I think it would be better to iterate on each `requestQueryParams.getAll(forwardedParam)` element and add that to `codeFlowParams` instead.

https://en.wikipedia.org/wiki/Query_string states that it is allowed to have multiple values to be associated with a single field (e.g. `field1=value1&field1=value2&field2=value3`)

Reply:
Hey George, this is what it does I think, `requestQueryParams.getAll(forwardedParam)` will return all values for a given param, example, `field1=value1&field1=value2` will return this List of 2 values - but I'll update the test to verify

Reply:
Or you mean pass them literally like they were provided ? Sure, I can do that

Reply:
> Or you mean pass them literally like they were provided ? Sure, I can do that

Yes, pass them as they were provided

Reply:
`Field1=a&Field1=b` instead of `Field1=a%20b`

Reply:
@viswatejan AFAIU those parameters are not passed along unless they are specified in the `application.properties`

Reply:
@gastaldi, you are correct only the ones specified in the config will be forwarded and that should also be stripped off from the state cookie. You can check the `generateCodeFlowState` method and it is actually adding query params to the state cookie and, that path is what is being used after the identity provider redirects to the application with code query param. So stripping those forwardable query params before storing them in the state cookie will make sure the final application URL that the user sees after successful authentication will not have those query params.

Or we can also add additional configuration property to control this behavior(exclude or include it in the final application URL).

Reply:
@viswatejan I think it will complicate things, these parameters are part of the initial request URL,  they are not coming from Keycloak but provided to Quarkus in the request URL, therefore the requester is not concerned about the query parameters entering Quarkus URI space. Quarkus application does not need to use them or be aware of them...

Reply:
If I understand the current implementation, the following is what happens:

Requested Quarkus app URL
`https://my-app.com?kc_idp_hint=google&application_query_param=welcome`
> Assumption is only `kc_idp_hint` is configured to forward to the Keycloak and not the `application_query_param`.

`quarkus-oidc` library will store the requested quarkus app URL along with query params in the state cookie and build the authorization code flow URL with just the app URL without the query params as the redirect_uri like below:
`https://keycloak-idp-.com/auth/realms/my-realm/protocol/openid-connect/auth?client_id=my-client&redirect_uri=https://my-app.com&response_type=code&scope=openid&response_mode=fragment&kc_idp_hint=google&state=xyz`

After successful authentication, the user will be redirected to
`https://my-app.com?code=1234&state=xyz`

And the `quarkus-oidc` library intercepts and refer to the state cookie and makes a final redirect to 
`https://my-app.com?kc_idp_hint=google&application_query_param=hi`
but the expectation is 
`https://my-app.com?application_query_param=welcome`

So, I was suggesting to remove the forwarded params from the state cookie value before saving the cookie or at least remove the forwarded params in the final redirect.

Reply:
@viswatejan 

So, given the final `https://my-app.com?kc_idp_hint=google&application_query_param=welcome`, why do the extra config/code to remove `kc_idp_hint=google` ? It is not a sensitive info, it was part of the original request ?

Reply:
@sberyozkin, `kc_idp_hint` may not have the sensitive information but `login_hint`(can contain email) or any other custom forwardable params might have sensitive/PII information and that is the only reason, I was thinking to keep the final application URL clean from all the forwarded query params.

If the consumers of this extension need those forwarded query params to be not present on the final application URL, they have to code for additional redirect with those forwarded query params removed and so I was thinking it will be better to implement it in the extension itself.

I am fine with your current changes if you think implementing this will complicate or confuse the developers.

Reply:
@viswatejan if the email was already in the forwarded params and then got redirected via the browser to KC then it has already been potentially leaked.
That said, I agree it is cleaner not to have such forwarded params included in the final URI, for cleaner URI links, etc, so I updated the code a bit to strip them

Reply:
@gastaldi I've updated the code to make sure all the param values are included, the only problem Keycloak does not like things like `kc_idp_hint=google&kc_idp_hint=azure` so I did not test that multi-value query params are forwarded, only a single value; the code for dealing with multiple values is quite simple though